enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1054 @@
+
+CoW is a feature of filesystem (supported by XFS and Btrfs) and it can be used by 
+providing a `--reflink=` param to `cp` option.
+`--reflink=auto` will be used over `--reflink=always` to gracefully fall back to regular
+copying on filesystems not supporting CoW (ext4, ZFS).
+Since CoW is backed by filesystem, it works only within that filesystem.
+
+To keep track of which backup is intended for which deployment, backup will be placed in a
+directory named after ostree deployment ID inside `/var/lib/microshift.bak/` dir, e.g.
+`/var/lib/microshift.bak/rhel-8497faf62210000ffb5274c8fb159512fd6b9074857ad46820daa1980842d889.0`.
+
+If MicroShift starts and notices that data directory exists, metadata doesn't exist,
+`next-boot` file is missing, it will be assumed that it's a 4.13 to 4.14 upgrade.
+In such case ID of rollback deployment will be used but the automatic rollback will not
+work as 4.13 will not include restore capabilities.
+
+Additionally, a symlink `/var/lib/microshift.bak/latest` will point to the most recent
+backup. This will allow redboots of staged deployment to restore backup of previous
+deployment to use as a starting point (to protect and retry in case of failed data
+migration).
+
+Restore operation works the same, just in the other direction - copying contents of 
+`/var/lib/microshift.bak/ostree-deploy-id/` to `/var/lib/microshift/`.
+
+End user documentation needs to include:
+- guidance on setting up filesystem to fullfil requirements for using copy-on-write
+  (e.g. making sure some filesystem options are not disabled).
+- remark that in case of missing CoW support, full backup will be made.
+
+### Contents of MicroShift data backup
+
+- etcd database shall be backed up fully, rather just creating a snapshot to preserve
+  history and other data that could be lost.
+- Kubeconfigs and certificates needs to be backup and restored in order to keep communication working.
+  MicroShift could regenerate them, but it would mean that kubeconfigs used to remotely access cluster
+  will become invalid.
+  - Following approach might mean that certificates might need to be amended to update the
+    Subject Alternative Names (SAN) list.
+- Versions of binaries don't impact decision whether to perform snapshot or copy whole data
+  directory because "newer" version will need to read the existing data anyway to
+  perform the data migration.
+
+Based on reasons above, it was decided that whole `/var/lib/microshift` will be backed up.
+
+### MicroShift version metadata persistence
+
+When MicroShift is up and running healthy, green script will persist version into a file 
+within data dir, e.g. `/var/lib/microshift/version`:
+```plaintext
+4.14.0
+```
+
+Value will be used during data migration to decide if
+- migration can be skipped,
+- migration needs to be blocked,
+- migration can be attempted.
+
+Creating the file by green script will indicate what was the MicroShift version that
+successfully ran using the data.
```

Comment:
We may want to consider keeping a history/log of the changes, rather than a 1 time file. That will help with support understanding what has happened to the data directory and it may make it easier to detect some of the edge cases.

Reply:
Included small section `### Action log`